src/pudl/io_managers.py

```diff
@@ -381,13 +369,45 @@ def _create_geoparquet_metadata(self, gdf: gpd.GeoDataFrame, res: Resource) -> s
             "primary_column": primary_column,
             "columns": geometry_columns,
         }
+        return json.dumps(geo_metadata).encode()
 
-        return json.dumps(geo_metadata)
-
-    def load_input(self, context: InputContext) -> pd.DataFrame | gpd.GeoDataFrame:
-        """Loads pudl table from parquet file."""
+    def handle_output(self, context: OutputContext, obj: gpd.GeoDataFrame) -> None:
+        """Write a PUDL dataframe to GeoParquet."""
+        if not isinstance(obj, gpd.GeoDataFrame):
+            raise TypeError(
+                f"Only geopandas dataframes are supported, got {type(obj)}."
+            )
         table_name = get_table_name_from_context(context)
-        return get_parquet_table(table_name)
+        parquet_path = PudlPaths().parquet_path(table_name)
+        parquet_path.parent.mkdir(parents=True, exist_ok=True)
+
+        res = Resource.from_id(table_name)
+        gdf = res.enforce_schema(obj)
+
+        # Extract metadata before modifying geometry columns
+        geo_metadata = self._create_geoparquet_metadata(gdf, res)
+        # Convert geometry columns to WKB
+        geometry_fields = [
+            field.name
+            for field in res.schema.fields
+            if field.type == "geometry" and field.name in gdf.columns
+        ]
+        for field_name in geometry_fields:
+            # This conversion is required to get the right data into the Parquet output
+            # but it isn't technically compatible with the GeoDataFrame, so we get a
+            # warning from Geopandas about the geometry column not being a geometry.
+            logger.info(f"Convert geometry column {table_name}.{field_name} to WKB.")
+            gdf[field_name] = gdf[field_name].to_wkb()
+
+        # Convert to PyArrow table with explicit schema
+        pa_table = pa.Table.from_pandas(
+            gdf, schema=res.to_pyarrow(), preserve_index=False
+        )
+        # Add GeoParquet metadata
+        metadata = pa_table.schema.metadata or {}
```

Comment:
non-blocking Q: does `pa_table.schema.metadata` have to have `bytes` instead of `str` everywhere?

Reply:
Actually it seems like this byte encoding is entirely unnecessary. [Just leaving it a JSON string](https://geoparquet.org/releases/v1.0.0/).

> A GeoParquet file MUST include a geo key in the Parquet metadata (see [FileMetaData::key_value_metadata](https://github.com/apache/parquet-format#metadata)). The value of this key MUST be a JSON-encoded UTF-8 string representing the file and column metadata that validates against the [GeoParquet metadata schema](https://geoparquet.org/releases/v1.0.0/schema.json). The file and column metadata fields are described below.